### ISSUE TITLE
Commander exploding for no reason / Looks like CTRL+K, but not ?

### ISSUE THREAD

**Issue by MechSanGoku** [FAForever/fa#3406]
Hi All,

For some reason, around min 10 of the game, I had the message:
MSG has CTRL+K

Weird thing is, and I was vocal with teammates - They saw this too, I replied I didn't die, and my com was still here.

My com was still alive for a few seconds, and boom, it exploded, just like a CTRL+K a few seconds later (possibly 5 ?)
But I never did press that. I even have the self destruct timer per default
https://ibb.co/ygFJSt7

https://replay.faforever.com/15261648

Watching the replay, this can be observed around 10:30 / 10:45

Thread on FAF Forum:
https://forum.faforever.com/topic/2511/com-exploding-for-no-reason-wasn-t-ctrl-k?_=1631030436572

**Comment by baurse** [comment]
Jup, is weird. 
It looks like a typical disconnect, but obviously isn't one, and the 5 second ctrl+k countdown is not shown in the game, so a missclick can probably be excluded as well.

**Comment by Garanas** [comment]
The good news is that it wasn't on FAF Develop, the bad news is that it is likely also happening on FAF Develop.

I recall this happened before - lets use this issue as a place to find all of them and see how they connect.

**Comment by KionX** [comment]
Evildrew:
hi Gieb, don't know if you are asleep but anyway,
can you pass on to the technical people this replay #15221176 at about 21-22 mins in I am defeated for no apparent reason,
then a few seconds later my acu explodes so isnt an accidental ctrl k,
i wrote in chat after i was defeated so i didnt leave, didnt make any sense...

**Comment by KionX** [comment]
It died because GetFractionComplete returned a value less than one.
Someone disassembled it within a few ticks. I was not able to fully reproduce this.
A negative build rate was needed.

**Comment by KionX** [comment]
15261648 and 15221176 died because GetFractionComplete returned a value less than one.

**Comment by Garanas** [comment]
In both replays the commander was being repaired by engineers. What is the value that is being returned? Could it be a rounding issue of some sort?

**Comment by Garanas** [comment]
Could it be because of the player is stalling?

**Comment by KionX** [comment]
This problem probably applies to all repairs and may be construction. But only on the ACU it is noticeable.
I'll try to look deeper later.

**Comment by baurse** [comment]
Figured out another hint: 
Both players loose an energy storage and have a spike in energy usage at the exact moment the ctrl+k message appears.

MSG got gifted the base by his ally, which explodes a couple seconds later, generating the energy spike and also destroying the storage.
Evildrew has an energy spike for seemingly random reasons (overflow cutting out?) while reclaiming one of his e-storages.

When watching the replay at -10 from 10:35 or 20:37 respectively and staring at the energy bar the pattern is pretty clear:
Energy usage spike + loose 5k energy storage => Ctrl+K Message appears.

Not saying that the above causes the bug, but it seems related at least.

Sadly I couldn't reproduce the bug by just going into a test game, repairing my damaged ACU and causig an energy spike + energy storage change by blowing up some pgens and storages. Maybe overflow is related in some way?

**Comment by Garanas** [comment]
Good finds! Lets hope (or not) that we get one or two additional replays where this happens.

**Comment by KionX** [comment]
Similar: 15175485
GetEconomyTotals() at this moment outputs
```
income={ ENERGY=787.14038085938, MASS=-55838.76953125 },
lastUseActual={ ENERGY=-1742218.125, MASS=-40338.703125 }, //expense
lastUseRequested={ ENERGY=808.43078613281, MASS=14.780558586121 },
maxStorage={ ENERGY=34000, MASS=1540 },
reclaimed={ ENERGY=31694.900390625, MASS=16390.775390625 },
stored={ ENERGY=34000, MASS=1540 }
```

**Comment by FemtoZetta** [comment]
Might be related, replay #15313685 at like 7:40min or so, green mid starts T2 upgrade and dies shortly after.

**Comment by KionX** [comment]
> Might be related, replay [#15313685](https://replay.faforever.com/15313685) at like 7:40min or so, green mid starts T2 upgrade and dies shortly after.

This is a normal Ctrl-K

**Comment by FemtoZetta** [comment]
Ah you are right, I didn't pay close enough attention. Nvm then.

**Comment by KionX** [comment]
Intermediate result for 15261648
On the tick 6342 energy income of the ARMY_10 became -9954.
Therefore, the engineers received a negative build rate and disassembled the ACU.

There was another moment: 6230 ARMY_8 -7946.66

**Comment by KionX** [comment]
The problem was reproduced.
You need to have an ally who shares resources, destroy the used energy storage, income is less than losses, and something else...
So income becomes antimatter.

Fixed [ForgedAlliance_mod.zip](https://github.com/FAForever/fa/files/7172640/ForgedAlliance_mod.zip)
Replaced negative values with zero.

**Comment by Garanas** [comment]
Great work - I'll see if I can reproduce it and look at the other issue (about gifting) too this weekend.

**Comment by KionX** [comment]
We need to decide whether we want to include fixes for this and #3235 in the patch.

**Comment by Garanas** [comment]
Fixed with the executable available on FAF Develop, closing it.
